### ISSUE TITLE
3D Water Cup waterLevelProperty not setting state appropriately

### ISSUE THREAD

**Issue by marlitas** [phetsims/mean-share-and-balance#21]
In phetio state wrapper the downstream 3D water cup is not tracking the waterLevelProperty state of the upstream cup.

**Comment by marlitas** [comment]
@zepumph Patch for PhetioStateEngine error message changes. Prettified error message display. 
<details>

```diff
Index: js/PhetioStateEngine.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/PhetioStateEngine.js b/js/PhetioStateEngine.js
--- a/js/PhetioStateEngine.js	(revision bb047fff83b3010c0dd5fe8a31e6b42cafec6054)
+++ b/js/PhetioStateEngine.js	(date 1653331576994)
@@ -327,15 +327,14 @@
 
         const unsetState = {};
         nextPass.forEach( phetioID => {
-          unsetState[ phetioID ] = state[ phetioID ];
-        } );
-
-        const assertMessage = `Impossible set state from iterate; unset state:\n ${JSON.stringify( unsetState, null, 2 )}`;
-        assert && assert( false, assertMessage );
+        unsetState[ phetioID ] = state[ phetioID ];
+      } );
+        const assertMessageTitle = 'Impossible set state from iterate; unset state:';
+        assert && assert( false, assertMessageTitle, unsetState );
 
         // We must exit here even if assertions are disabled so it wouldn't lock up the browser.
         if ( !assert ) {
-          throw new Error( assertMessage );
+          throw new Error( `${assertMessageTitle}\n ${JSON.stringify( unsetState, null, 2 )}` );
         }
       }
     }
```

**Comment by samreid** [comment]
Paper trail from my and @marlitas investigation:

<details>

```diff

Index: main/mean-share-and-balance/js/common/MeanShareAndBalanceConstants.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/mean-share-and-balance/js/common/MeanShareAndBalanceConstants.ts b/main/mean-share-and-balance/js/common/MeanShareAndBalanceConstants.ts
--- a/main/mean-share-and-balance/js/common/MeanShareAndBalanceConstants.ts	(revision 563c477ae70c3d955e1c6ca8539f19b6ca6563f5)
+++ b/main/mean-share-and-balance/js/common/MeanShareAndBalanceConstants.ts	(date 1653406112716)
@@ -18,7 +18,7 @@
   INITIAL_NUMBER_OF_CUPS: 1,
   CUP_WIDTH: 50,
   CUP_HEIGHT: 100,
-  WATER_LEVEL_DEFAULT: 0.5,
+  WATER_LEVEL_DEFAULT: 0.25,
   PIPE_STROKE_WIDTH: 1,
   PIPE_LENGTH: 50,
   CONTROLS_VERTICAL_MARGIN: 30,
Index: main/sun/js/Slider.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/sun/js/Slider.ts b/main/sun/js/Slider.ts
--- a/main/sun/js/Slider.ts	(revision 872f33b097e1195c719f42af189c37050857bd6d)
+++ b/main/sun/js/Slider.ts	(date 1653409039644)
@@ -445,7 +445,9 @@
     const enabledRangeObserver = function( enabledRange: Range ) {
 
       // clamp the value to the enabled range if it changes
-      valueProperty.set( Utils.clamp( valueProperty.value, enabledRange.min, enabledRange.max ) );
+      if ( !phet.joist.sim.isSettingPhetioStateProperty.value ) {
+        valueProperty.set( Utils.clamp( valueProperty.value, enabledRange.min, enabledRange.max ) );
+      }
     };
     this.enabledRangeProperty.link( enabledRangeObserver ); // needs to be unlinked in dispose function
 
Index: main/axon/js/Property.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/axon/js/Property.ts b/main/axon/js/Property.ts
--- a/main/axon/js/Property.ts	(revision 8df827f6c089014d4a06ec8b8f4139960fc0b4b1)
+++ b/main/axon/js/Property.ts	(date 1653408206633)
@@ -196,6 +196,16 @@
    * hasn't changed, this is a no-op.
    */
   set( value: T ): void {
+
+    if ( this.tandem && this.tandem.phetioID.startsWith( 'meanShareAndBalance.levelingOutScreen.model.waterCup3DGroup.waterCup3D' ) ) {
+      console.log( phet.preloads.phetio.queryParameters.frameTitle + ': ' + value + ', this.isDeferred=' + this.isDeferred );
+
+      // @ts-ignore
+      if ( this.isDeferred && value === 0.25 ) {
+        debugger;
+      }
+    }
+
     if ( !this.isDisposed ) {
       if ( this.isDeferred ) {
         this.deferredValue = value;
@@ -301,6 +311,10 @@
   setDeferred( isDeferred: boolean ): ( () => void ) | null {
     assert && assert( !this.isDisposed, 'cannot defer Property if already disposed.' );
     assert && assert( typeof isDeferred === 'boolean', 'bad value for isDeferred' );
+
+    if ( this.tandem && this.tandem.phetioID.startsWith( 'meanShareAndBalance.levelingOutScreen.model.waterCup3DGroup.waterCup3D' ) ) {
+      console.log( phet.preloads.phetio.queryParameters.frameTitle + ': setDeferred: ' + isDeferred );
+    }
     if ( isDeferred ) {
       assert && assert( !this.isDeferred, 'Property already deferred' );
       this.isDeferred = true;
@@ -315,6 +329,7 @@
       if ( this.hasDeferredValue ) {
         this.setPropertyValue( this.deferredValue! );
         this.hasDeferredValue = false;
+        this.deferredValue = null;
       }
 
       // If the value has changed, prepare to send out notifications (after all other Properties in this transaction
Index: main/mean-share-and-balance/js/intro/model/WaterCupModel.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/mean-share-and-balance/js/intro/model/WaterCupModel.ts b/main/mean-share-and-balance/js/intro/model/WaterCupModel.ts
--- a/main/mean-share-and-balance/js/intro/model/WaterCupModel.ts	(revision 563c477ae70c3d955e1c6ca8539f19b6ca6563f5)
+++ b/main/mean-share-and-balance/js/intro/model/WaterCupModel.ts	(date 1653408045377)
@@ -48,6 +48,7 @@
       range: new Range( 0, 1 ),
       tandem: options.tandem.createTandem( 'waterLevelProperty' )
     }, options.waterLevelPropertyOptions ) );
+    console.log( phet.preloads.phetio.queryParameters.frameTitle + ': ' + this.waterLevelProperty.value );
   }
 
   override dispose(): void {
Index: main/mean-share-and-balance/js/intro/model/IntroModel.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/mean-share-and-balance/js/intro/model/IntroModel.ts b/main/mean-share-and-balance/js/intro/model/IntroModel.ts
--- a/main/mean-share-and-balance/js/intro/model/IntroModel.ts	(revision 563c477ae70c3d955e1c6ca8539f19b6ca6563f5)
+++ b/main/mean-share-and-balance/js/intro/model/IntroModel.ts	(date 1653405512658)
@@ -80,6 +80,7 @@
     } );
 
     this.waterCup2DGroup = new PhetioGroup( ( tandem: Tandem, x: number ) => {
+      console.log( phet.preloads.phetio.queryParameters.frameTitle + ': 2d cup group constructor callback: ' + tandem.phetioID );
       return new WaterCupModel( {
         tandem: tandem,
         x: x,
@@ -91,7 +92,8 @@
     }, [ 0 ], {
       phetioType: PhetioGroup.PhetioGroupIO( WaterCupModel.WaterCupModelIO ),
       phetioDocumentation: 'Holds the models for the 2D water cups.',
-      tandem: options.tandem.createTandem( 'waterCup2DGroup' )
+      tandem: options.tandem.createTandem( 'waterCup2DGroup' ),
+      supportsDynamicState: false
     } );
 
     this.pipeGroup = new PhetioGroup( ( tandem: Tandem, x: number, y: number ) => {
@@ -99,7 +101,8 @@
     }, [ 0, 0 ], {
       phetioType: PhetioGroup.PhetioGroupIO( PipeModel.PipeModelIO ),
       phetioDocumentation: 'Holds the connecting pipes for the 2D water cups.',
-      tandem: options.tandem.createTandem( 'pipeGroup' )
+      tandem: options.tandem.createTandem( 'pipeGroup' ),
+      supportsDynamicState: false
     } );
 
     this.meanProperty = new NumberProperty( MeanShareAndBalanceConstants.WATER_LEVEL_DEFAULT, {
@@ -111,54 +114,69 @@
 
     const validWaterLevelRange = new Range( 0, 1 );
 
-    // add/remove water cups and pipes according to number spinner
-    this.numberOfCupsProperty.link( numberOfCups => {
-      while ( numberOfCups > this.waterCup3DGroup.count ) {
-
-        const lastWaterCup: WaterCupModel | null = this.waterCup3DGroup.count > 0 ? this.waterCup3DGroup.getLastElement() : null;
-        const x = lastWaterCup ?
-                  lastWaterCup.x + ( MeanShareAndBalanceConstants.CUP_WIDTH + MeanShareAndBalanceConstants.PIPE_LENGTH ) :
-                  0;
-        const new3DCup = this.waterCup3DGroup.createNextElement( x );
-        const new2DCup = this.waterCup2DGroup.createNextElement( x );
-
-        // Wire up the water level, treating the 3d model as the ground truth
-        new3DCup.waterLevelProperty.lazyLink( ( waterLevel, oldWaterLevel ) => {
-
-          const delta = waterLevel - oldWaterLevel;
-          const new2DWaterLevel = new2DCup.waterLevelProperty.value + delta;
-          new2DCup.waterLevelProperty.value = validWaterLevelRange.constrainValue( new2DWaterLevel );
+    this.waterCup3DGroup.elementCreatedEmitter.addListener( waterCup3D => {
+      console.log( phet.preloads.phetio.queryParameters.frameTitle + ': 3d cup created' );
+      const lastWaterCup: WaterCupModel | null = this.waterCup3DGroup.count > 1 ? this.waterCup3DGroup.getElement( this.waterCup3DGroup.count - 2 ) : null;
+      const x = lastWaterCup ?
+                lastWaterCup.x + ( MeanShareAndBalanceConstants.CUP_WIDTH + MeanShareAndBalanceConstants.PIPE_LENGTH ) :
+                0;
+      const new2DCup = this.waterCup2DGroup.createCorrespondingGroupElement( waterCup3D.tandem.name, x );
+      console.log( phet.preloads.phetio.queryParameters.frameTitle + ': 2d cup created' );
+      waterCup3D.waterLevelProperty.lazyLink( ( waterLevel, oldWaterLevel ) => {
+        const delta = waterLevel - oldWaterLevel;
+        const new2DWaterLevel = new2DCup.waterLevelProperty.value + delta;
+        new2DCup.waterLevelProperty.value = validWaterLevelRange.constrainValue( new2DWaterLevel );
 
-          // Constrain waterHeightRange on 3D cup if corresponding 2D Cup runs out of space.
-          new2DCup.waterLevelProperty.value === 1 ?
-          new3DCup.waterHeightRange.setMax( new3DCup.waterLevelProperty.value ) :
-          new3DCup.waterHeightRange.setMax( 1 );
+        // Constrain waterHeightRange on 3D cup if corresponding 2D Cup runs out of space.
+        new2DCup.waterLevelProperty.value === 1 ?
+        waterCup3D.waterHeightRange.setMax( waterCup3D.waterLevelProperty.value ) :
+        waterCup3D.waterHeightRange.setMax( 1 );
 
-          // If 2D cup reaches 0 and 3D cup still has water
-          // Subtract delta from max waterLevelProperty in 2D cups
-          if ( new2DCup.waterLevelProperty.value === 0 ) {
-            let fullestCup = this.waterCup3DGroup.getElement( 0 );
-            this.waterCup2DGroup.forEach( cup => {
-              cup.waterLevelProperty.value > fullestCup.waterLevelProperty.value && ( fullestCup = cup );
-            } );
+        // If 2D cup reaches 0 and 3D cup still has water
+        // Subtract delta from max waterLevelProperty in 2D cups
+        if ( new2DCup.waterLevelProperty.value === 0 ) {
+          let fullestCup = this.waterCup3DGroup.getElement( 0 );
+          this.waterCup2DGroup.forEach( cup => {
+            cup.waterLevelProperty.value > fullestCup.waterLevelProperty.value && ( fullestCup = cup );
+          } );
 
-            fullestCup.waterLevelProperty.set( fullestCup.waterLevelProperty.value + delta );
-          }
-
-          this.updateMeanFrom3DCups();
-        } );
+          fullestCup.waterLevelProperty.set( fullestCup.waterLevelProperty.value + delta );
+        }
+        this.updateMeanFrom3DCups();
+      } );
+      this.updateMeanFrom3DCups();
+    } );
 
-        lastWaterCup && this.pipeGroup.createNextElement( lastWaterCup.x, new2DCup.y );
-      }
-      while ( numberOfCups < this.waterCup3DGroup.count ) {
-        this.waterCup3DGroup.disposeElement( this.waterCup3DGroup.getLastElement() );
-        this.waterCup2DGroup.disposeElement( this.waterCup2DGroup.getLastElement() );
-        this.matchCupWaterLevels();
-        if ( numberOfCups > 0 ) {
-          this.pipeGroup.disposeElement( this.pipeGroup.getLastElement() );
-        }
+    this.waterCup2DGroup.elementCreatedEmitter.addListener( waterCup2D => {
+      const lastWaterCup: WaterCupModel | null = this.waterCup2DGroup.count > 1 ? this.waterCup2DGroup.getElement( this.waterCup2DGroup.count - 2 ) : null;
+      lastWaterCup && this.pipeGroup.createNextElement( lastWaterCup.x, waterCup2D.y );
+    } );
+
+    // TODO: match up 3D cup index to 2D cup index
+    this.waterCup3DGroup.elementDisposedEmitter.addListener( waterCup3D => {
+      this.waterCup2DGroup.disposeElement( this.waterCup2DGroup.getLastElement() );
+      this.waterCup3DGroup.count > 0 && this.updateMeanFrom3DCups();
+    } );
+
+    this.waterCup2DGroup.elementDisposedEmitter.addListener( waterCup2D => {
+      if ( this.waterCup2DGroup.count > 1 ) {
+        this.pipeGroup.disposeElement( this.pipeGroup.getLastElement() );
+      }
+    } );
+
+    // add/remove water cups and pipes according to number spinner
+    this.numberOfCupsProperty.link( numberOfCups => {
+      while ( numberOfCups > this.waterCup3DGroup.count ) {
+
+        const lastWaterCup: WaterCupModel | null = this.waterCup3DGroup.count > 0 ? this.waterCup3DGroup.getLastElement() : null;
+        const x = lastWaterCup ?
+                  lastWaterCup.x + ( MeanShareAndBalanceConstants.CUP_WIDTH + MeanShareAndBalanceConstants.PIPE_LENGTH ) :
+                  0;
+        this.waterCup3DGroup.createNextElement( x );
       }
-      this.updateMeanFrom3DCups();
+      while ( numberOfCups < this.waterCup3DGroup.count ) {
+        this.waterCup3DGroup.disposeElement( this.waterCup3DGroup.getLastElement() );
+      }
 
       assert && assert( numberOfCups === this.waterCup3DGroup.count, `Expected ${numberOfCups} cups, but found: ${this.waterCup3DGroup.count}.` );
       assert && assert( numberOfCups > 0, 'There should always be at least 1 cup' );
Index: main/tandem/js/PhetioGroup.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/tandem/js/PhetioGroup.ts b/main/tandem/js/PhetioGroup.ts
--- a/main/tandem/js/PhetioGroup.ts	(revision 95b9294004da8027cb5321a1b98f9052961be8c1)
+++ b/main/tandem/js/PhetioGroup.ts	(date 1653407547023)
@@ -186,6 +186,8 @@
    * Remove and dispose all registered group elements
    */
   override clear( options?: any ): void {
+
+    console.log( 'clearing group!!! ' + this.tandem.phetioID );
     options = merge( {
 
       // used for validation during state setting (phet-io internal), see PhetioDynamicElementContainer.disposeElement for documentation
Index: main/scenery-phet/js/BeakerNode.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/scenery-phet/js/BeakerNode.ts b/main/scenery-phet/js/BeakerNode.ts
--- a/main/scenery-phet/js/BeakerNode.ts	(revision b4fc53a0e4d62d4c20357c9761b0c65c6f874b60)
+++ b/main/scenery-phet/js/BeakerNode.ts	(date 1653405972119)
@@ -138,8 +138,12 @@
       pickable: false
     } );
 
+    console.log( phet.preloads.phetio.queryParameters.frameTitle + ': ' + solutionLevelProperty.tandem.phetioID );
     // solution level adjustment listener
     solutionLevelProperty.link( solutionLevel => {
+
+      console.log( phet.preloads.phetio.queryParameters.frameTitle + ': ' + solutionLevel );
+
       const centerLiquidY = centerBottom - options.beakerHeight * solutionLevel;
       const solutionTopShape = new Shape()
         .ellipticalArc( 0, centerLiquidY, options.xRadius, options.yRadius, 0, 0, Math.PI * 2, false )
Index: main/phet-io/js/PhetioStateEngine.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/phet-io/js/PhetioStateEngine.js b/main/phet-io/js/PhetioStateEngine.js
--- a/main/phet-io/js/PhetioStateEngine.js	(revision 2089106e00fb9a37756a001440ad673427718833)
+++ b/main/phet-io/js/PhetioStateEngine.js	(date 1653406962931)
@@ -22,6 +22,7 @@
 import Tandem from '../../tandem/js/Tandem.js';
 import dataStream from './dataStream.js';
 import phetio from './phetio.js';
+import phetioEngine from './phetioEngine.js';
 
 // constants
 const DELETED_VALUE = 'DELETED'; // this is copied in documentation for PhetioEngineIO.getChangedState(), and used in Studio.
@@ -262,6 +263,11 @@
   setFullState( state ) {
     this.setState( state, Tandem.ROOT );
     this.mostRecentSavedState = state;
+
+    const waterLevelProperty = phetioEngine.getPhetioObject( 'meanShareAndBalance.levelingOutScreen.model.waterCup3DGroup.waterCup3D_0.waterLevelProperty' );
+    console.log( waterLevelProperty );
+    console.log( waterLevelProperty.isDeferred );
+    debugger;
   }
 
   /**
@@ -292,6 +298,10 @@
       try {
         let phetioObject = null;
 
+        // if ( phetioID === 'meanShareAndBalance.levelingOutScreen.model.waterCup3DGroup.waterCup3D_0.waterLevelProperty' ) {
+        //   debugger;
+        // }
+
         // If we already have a reference to the PhetioObject
         if ( this.phetioEngine.phetioObjectMap.hasOwnProperty( phetioID ) ) {
           phetioObject = this.phetioEngine.getPhetioObject( phetioID );
@@ -327,8 +337,8 @@
 
         const unsetState = {};
         nextPass.forEach( phetioID => {
-        unsetState[ phetioID ] = state[ phetioID ];
-      } );
+          unsetState[ phetioID ] = state[ phetioID ];
+        } );
         const assertMessageTitle = 'Impossible set state from iterate; unset state:';
         assert && assert( false, assertMessageTitle, unsetState );
 
```

</details>

Using this patch, we found that Slider is setting the deferred `waterLevelProperty` value to its current value as part of `clamp`.  This means the deferred property thinks it should take the old value when undeferred.  We will commit a guard like `!isSettingStateProperty.value` and request discussion with @zepumph .

**Comment by marlitas** [comment]
Journey to reverting commit: 2e4a346

Commit in question shifted the responsibility of dynamic element creation and disposal from the numberSpinner to phetioGroups. This shift was made through the suggestion of @zepumph to properly couple dynamic element creation rather than relying on a NumberProperty to handle all creation and disposal. Element relationships were made clearer through this, and resulted in:
 - waterCup3D elements created inside a NumberProperty link
 - waterCup2D elements created inside a waterCup3D emitter listener
 - pipe elements created inside a waterCup2D emitter listener
 
 The same pattern was followed for disposal.  

After battling several phetio state wrapper errors @samreid and I decided to move responsibility of dynamic element creation back solely to the NumberProperty. 

The errors we encountered were:
- "Impossible set state from iterate; unset state:" (waterCup2DGroup, pipeGroup)
  - Fixed waterCup2DGroup by creating elements with: ```createCorrespondingGroupElement```
- Attempted same solution with pipeGroup, however waterCup2DGroup would remain empty when elementCreatedEmitter fired, therefore the pipeGroup had no accurate reference for creation.

We decided to revert back to NumberProperty as the responsible party in element creation and disposal due to the simplicity when dealing with state. If phetioGroup defferals, emitters, and creation order can be clarified we may choose to implement the reverted commit again.
